File: Dynamic Programming/Python Algorithms/Codility/Problem Algorithms/Painless/equileader.py
#Codility EquiLeader problem


#Poor Solution **44%**
def equileader(A):
    def dominator_helper(B):
        M = len(B)
        size = 0
        value = 0
        count = 0
        candidate = 0
        if (M == 2):
            if (B[0] != B[1]): return 0
            else: return B[0]

        for i in range(M):
            if (size == 0):
                size += 1
                value = B[i]
            elif (B[i] != value):
                size -= 1
            else:   size += 1
        
        if (size > 0): candidate = value
        for i in range(M):
            if (B[i] == candidate):
                count += 1
        if (count > M//2):  return candidate
        else:   return 0

    N = len(A)
    if (N == 1): return 0
    if (N == 2):
            if (A[0] != A[1]): return 0
            else: return 1
    count_EquiLeader = 0
    for i in range(N - 1):
        K = [A[num] for num in range(N) if (num >= 0 and num <= i)]
        L = [A[num] for num in range(N) if (num > i and num < N)]
        
        Q = dominator_helper(K)
        R = dominator_helper(L)
        # print(K, L)
        # print(Q, R)
        if (( Q != 0 or R != 0) and Q == R):
            count_EquiLeader += 1
    return count_EquiLeader


  


print(equileader([4, 3, 4, 4, 4, 2]))
print(equileader([1, 2]))
print(equileader([1, 2, 3, 4, 5]))
print(equileader([1, 2, 2, 1, 2, 2]))
print(equileader([1,2,1,2,1,2,1,2,3,1]))

File: Dynamic Programming/JavaScript Algorithms/Codility/Problem Algorithms/Painless/equiLeader.js
//EquiLeader Codility Problem

//Find the index S such that the leaders of the sequences A[0], 
//A[1], ..., A[S] and A[S + 1], A[S + 2], ..., A[N - 1] are the same.


//Good Solution ***100%****
const equiLeader = (A) => {
    const N = A.length 

    let index = value = candidate = count = leader = k = 0
    const PrefLeader = Array(N).fill(0)


    for (let i = 0; i < N; i += 1){
        if (index === 0){
            index = 1
            value = A[i]
        } else if (value != A[i]){
            index -= 1
        } else index+= 1
    }
    if (index > 0) candidate  = value

    for (let num of A){
        if (candidate === num) count += 1
    }
    if (count > Math.floor(N/2)) {
        leader = candidate
        for (let i = 0; i < N; i += 1){
            if (A[i] === leader) PrefLeader[i] = k + 1
            else PrefLeader[i] = k
            k = PrefLeader[i]
        }
    }
    //console.log(leader, PrefLeader)
    let m = n = result = 0
    for (let i = 0; i < N; i += 1){
        let q = N - 1
        m = PrefLeader[i]
        n = PrefLeader[q]
        if (m > Math.floor((i + 1)/2) && (n - m) > Math.floor((q - i)/2)){
            result += 1
        }
    }
    return result  
}

console.log(equiLeader([4, 3, 4, 4, 4, 2]))
console.log(equiLeader([4, 4, 4, 4, 4, 4, 4, 4, 4]))
console.log(equiLeader([4, 3, 5, 5, 6, 8, 5, 5, -1, 6, 5, 5, 8, 5, 5]))
console.log(equiLeader([1, 0, 3, 1, 1, 1, 5, 1, 2, 1, 1, 1]))